### PR TITLE
Getting canvas usage data from Canvas analytics API

### LINKED ISSUE #25: Integrate canvas usage data from Analytics API
https://umich.instructure.com/api/v1/courses/course_id/analytics/activity

### PR DESCRIPTION
**Pull request by pushyamig**
Fixes #25 
Pulling Canvas usage data from the Canvas Analytics API
1. I am currently making a parallel call for pulling the data takes about ~5minutes(sequential calling take about 30 minutes) and generates about 500k records. I am deleting all the records and updating them with new data. Analytics API doesn't have a filter parameter by date for just query incremental approach. The deleting and insert 500k records went fine locally. 
2. I am retrying for a 500 status code. I see a lot of 404 from the canvas when ran the process, what that tells me is the data I am requesting for a course doesn't exist in the data source canvas is pulling from. I am not going to retry for those. I am having my workers defaults to 8 since I don't feel I need to increase more than that. I didn't see any rate-limiting during the most used canvas time.

### CODE REVIEW

File: canvas/canvas_usage.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+        if response is None:
+            logger.info(f"For Canvas usage response is None ")
+            return
+
+        logger.info(f"CanvasUsage date collected so far : {len(self.canvas_usage_courses)}")
+        status = response.result().status_code
+
+        course_id = response.result().url.split('courses/')[1].split('/')[0]
+
+        status_codes_to_retry = [500]
```
Comment (pushyamig) [inline canvas/canvas_usage.py:32]: If you want to test how my retry is working add 404 to this list. it will try for 3 times and moves on

File: canvas/published_date.py [hunk 2]
```diff
@@ -58,7 +58,7 @@ def published_date_resp_parsing(self, response):
         try:
             audit_events = json.loads(response.result().text)
         except JSONDecodeError as e:
-            logger.info(f"Error in parsing the response {e.message}")
+            logger.info(f"Error in parsing the response {e.msg}")
```
Comment (ssciolla) [inline canvas/published_date.py:61]: I guess this was a bug you wanted to fix? Might want to just make that change in a tiny separate PR, in case it comes up in production.

File: db/migrations/0005.canvas_usage_table.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+#
+# file: migrations/0005.canvas_usage_table.py
+#
+from yoyo import step
+
+__depends__ = {'0004.add_sis_id'}
+step('''
+    CREATE TABLE IF NOT EXISTS canvas_usage
+    (
+        canvas_id INTEGER NOT NULL,
```
Comment (ssciolla) [inline db/migrations/0005.canvas_usage_table.py:10]: Is this `canvas_id` for a user? I think `user_id` might be clearer then? It might also make sense to add an auto-incremented primary key.

File: canvas/canvas_usage.py [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        logger.info(f'Duration of Canvas usage run took: {str_time}')
+        return self.canvas_usage_to_df()
+
+
+
+
+
+
+
+
```
Comment (ssciolla) [inline canvas/canvas_usage.py:114]: Can you delete these please?

File: inventory.py [hunk 5]
```diff
@@ -237,6 +242,8 @@ def run_course_inventory() -> None:
         logger.info(f'Writing {num_enrollment_records} enrollment records to CSV')
         enrollment_df.to_csv(os.path.join('data', 'enrollment.csv'), index=False)
         logger.info('Wrote data to data/enrollment.csv')
+        logger.info(f"Writing {num_canvas_usage_records} canvas usage record to CSV")
+        canvas_usage_df.to_csv(os.path.join('data', 'canvas_usage_trails.csv'), index=False)
```
Comment (ssciolla) [inline inventory.py:246]: I'm curious, what is the `trails` part supposed to mean?

File: inventory.py [hunk 6]
```diff
@@ -257,6 +264,8 @@ def run_course_inventory() -> None:
     logger.info(f'Inserting {num_enrollment_records} enrollment records to DB')
     enrollment_df.to_sql('enrollment', db_creator_obj.engine, if_exists='append', index=False)
     logger.info(f'Inserted data into enrollment table in {db_creator_obj.db_name}')
+    logger.info(f"Inserting {num_canvas_usage_records} canvas_usage records to DB")
+    canvas_usage_df.to_sql('canvas_usage', db_creator_obj.engine, if_exists='append', index=False)
```
Comment (ssciolla) [inline inventory.py:268]: Might want to add the `Wrote` and `Inserted` after log messages for consistency. It's pretty wordy, but it's useful if there's an insertion problem.

File: canvas/published_date.py (hunk 2 above)
Comment (pushyamig) [reply canvas/published_date.py:61]: I just submitted the PR

File: db/migrations/0005.canvas_usage_table.py (hunk 3 above)
Comment (pushyamig) [reply db/migrations/0005.canvas_usage_table.py:10]: This is not at the user level but at the course level. This is course_id actually

File: canvas/canvas_usage.py (hunk 4 above)
Comment (pushyamig) [reply canvas/canvas_usage.py]: Done

File: inventory.py (hunk 5 above)
Comment (pushyamig) [reply inventory.py]: Removed it. I named it something random

File: canvas/published_date.py (hunk 2 above)
Comment (ssciolla) [reply canvas/published_date.py:61]: Yep, looks like we're good. I will re-build with this after the current job finishes, so the morning job will have it.

File: db/migrations/0005.canvas_usage_table.py (hunk 3 above)
Comment (ssciolla) [reply db/migrations/0005.canvas_usage_table.py:10]: Oh, gotcha. So maybe `course_id`?

Comment (pushyamig) [reply db/migrations/0005.canvas_usage_table.py:10]: Maybe I will change the table name to `course_canvas_usage` that will be more detailed to say. maybe in future we will track student-level who knows

File: inventory.py (hunk 6 above)
Comment (pushyamig) [reply inventory.py:269]: Done

File: inventory.py (hunk 5 above)
Comment (pushyamig) [reply inventory.py]: Done

**Comment by jennlove-um** [comment]
just as an fyi - here's the Canvas doc on what page views and participations mean in the context of the analytics: https://community.canvaslms.com/docs/DOC-16681-analytics-page-views-and-participations#jive_content_id_Overview

We'll pull the text from that for explaining what the data means.

File: canvas/canvas_course_usage.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        self.canvas_usage_courses = []
+        self.course_retry_list = []
+        self.retry_count = 0
+
+    def parsing_canvas_course_usage_data(self, response):
+        logger.info("parsing_canvas_course_usage_data Call")
+        new_dic = {}
+        if response is None:
+            logger.info(f"For Canvas Course usage response is None ")
+            return
```
Comment (ssciolla) [inline canvas/canvas_course_usage.py:26]: Nitpick: Can you return `None` when you do this? Seems a little clearer to me. I'm not insisting on you making the change.

**Comment by pushyamig** [comment]
_ means private access i.e in-class access only. You don't want the
instance of the class calling it outside of the class. It doesn't need to
be like that, I just chose it

File: canvas/canvas_course_usage.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+                logger.info("Append to retry list")
+                self.course_retry_list.append(course_id)
+            logger.info(f"Response not successful with status code {status} due to {response.result().text}")
+            return
+
+        try:
+            analytics_data = json.loads(response.result().text)
+        except JSONDecodeError as e:
+            logger.error(f"Error in parsing the response due to {e.msg}")
+            return
```
Comment (ssciolla) [inline canvas/canvas_course_usage.py:46]: I think you could also put this in the re-try mechanism, right? Maybe use a flag, something like what I did in my async PR with `problem_encountered`? I think at some point we may want to refactor so this mechanism is the same in all the classes we've written.

File: canvas/canvas_course_usage.py [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+
+        if not analytics_data:
+            logger.info(f"Response for fetching canvas course usage is empty {analytics_data}")
```
Comment (ssciolla) [inline canvas/canvas_course_usage.py:49]: Do we need to log `analytics_data` here if it's empty?

File: canvas/canvas_course_usage.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+            return
+        new_dic['course_id'] = course_id
+        new_dic['analytics'] = analytics_data
+
+        self.canvas_usage_courses.append(new_dic)
+
+    def _get_canvas_course_views_participation_data(self, retry_courses=None):
+        logger.info("Starting of _get_canvas_course_views_participation_data call")
+        with FuturesSession() as session:
+            headers = {'Content-type': 'application/json', 'Authorization': 'Bearer ' + self.canvas_token[0]}
```
Comment (ssciolla) [inline canvas/canvas_course_usage.py:59]: Can you explain what the indexing of `canvas_token` is doing here? It worked in your other PR, but don't you want the whole token? I must be missing something.

File: inventory.py [hunk 11]
```diff
@@ -14,6 +14,8 @@
 from canvas.published_date import FetchPublishedDate
 from canvas.async_enroll_gatherer import AsyncEnrollGatherer
 from gql_queries import queries as QUERIES
+from canvas.canvas_course_usage import CanvasCourseUsage
+
```
Comment (ssciolla) [inline inventory.py:18]:
```suggestion
from canvas.canvas_course_usage import CanvasCourseUsage
```

Comment (ssciolla) [reply inventory.py:18]: Just trying to remove this newline, not sure if it worked.

**Comment by pushyamig** [comment]
I felt I really don't want to retry there if there is some data specific
error. Retying may not resolve it. I felt that is overusing of retry
mechanism

File: inventory.py [hunk 12]
```diff
@@ -191,6 +193,10 @@ def run_course_inventory() -> None:
     course_df['published_at'] = pd.to_datetime(course_df['published_at'],
                                                format="%Y-%m-%dT%H:%M:%SZ",
                                                errors='coerce')
+    logger.info("*** Fetching the canvas course usage data ****")
+
```
Comment (ssciolla) [inline inventory.py:197]:
```suggestion
                                               errors='coerce')
   
    logger.info("*** Fetching the canvas course usage data ****")
```

**Comment by pushyamig** [comment]
It is sending a tuple like (2323232323,) so I just pulled the first one

**Comment by pushyamig** [comment]
Done

**Comment by pushyamig** [comment]
Oops, that was not needed. Removed it

**Comment by ssciolla** [comment]
> I felt I really don't want to retry there if there is some data specific error. Retying may not resolve it. I felt that is overusing of retry mechanism

Okay, I'm asking because this is specifically something I DID want to re-try for in my other Canvas calls. Sometimes the response is just invalid JSON, but the next attempt is fine. See `make_request_using_api_utils` in `inventory.py`

File: canvas/canvas_course_usage.py [hunk 13]
```diff
@@ -0,0 +1,107 @@
...
+
+        if not analytics_data:
+            logger.info(f"Response for fetching canvas course usage is empty {analytics_data}")
+            return
+        new_dic['course_id'] = course_id
+        new_dic['analytics'] = analytics_data
+
+        self.canvas_usage_courses.append(new_dic)
+
+    def _get_canvas_course_views_participation_data(self, retry_courses=None):
```
[the remark this answers is not in the document]
Comment (pushyamig) [reply canvas/canvas_course_usage.py:56]: https://github.com/tl-its-umich-edu/course-inventory/pull/51#issuecomment-608619633

File: canvas/canvas_course_usage.py [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+        self.retry_count = 0
+
+    def parsing_canvas_course_usage_data(self, response):
+        logger.info("parsing_canvas_course_usage_data Call")
+        new_dic = {}
+        if response is None:
+            logger.info(f"For Canvas Course usage response is None ")
+            return
+
+        logger.info(f"CanvasCourseUsage date collected so far : {len(self.canvas_usage_courses)}")
```
[the remark this answers is not in the document]
Comment (pushyamig) [reply canvas/canvas_course_usage.py]: Done

File: canvas/canvas_course_usage.py (hunk 8 above)
Comment (pushyamig) [reply canvas/canvas_course_usage.py:46]: https://github.com/tl-its-umich-edu/course-inventory/pull/51#issuecomment-608623356

File: canvas/canvas_course_usage.py (hunk 9 above)
Comment (pushyamig) [reply canvas/canvas_course_usage.py]: Done

File: canvas/canvas_course_usage.py (hunk 10 above)
Comment (pushyamig) [reply canvas/canvas_course_usage.py:59]: https://github.com/tl-its-umich-edu/course-inventory/pull/51#issuecomment-608624234

File: canvas/canvas_course_usage.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (ssciolla) [reply canvas/canvas_course_usage.py:56]: Replying to your comment below (the way email comments work in GitHub is strange), okay, yeah, the private and public designations I've used in Java. Just curious. The `_` is a little strange if it doesn't actually do anything in Python; there are probably decorators for this kind of thing?

File: canvas/canvas_course_usage.py (hunk 10 above)
Comment (ssciolla) [reply canvas/canvas_course_usage.py:59]: I don't understand your comment. Isn't `canvas_token` the personal access token string for Canvas?

File: canvas/canvas_course_usage.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+
+class CanvasCourseUsage:
+    def __init__(self, canvas_url, canvas_token, retry_attempts, course_ids):
+        self.canvas_url = canvas_url,
+        self.canvas_token = canvas_token,
+        self.course_ids = course_ids
+        self.retry_attempts = retry_attempts
+        self.canvas_usage_courses = []
+        self.course_retry_list = []
+        self.retry_count = 0
```
Comment (ssciolla) [inline canvas/canvas_course_usage.py:19]: Is this `retry_count` shared by all requests? Is it ever re-set? I think with the asynchronous approach, you need a dictionary like this:

```
{
   course_id: retry_int
}
```

File: canvas/canvas_course_usage.py (hunk 8 above)
Comment (ssciolla) [reply canvas/canvas_course_usage.py:46]: Please see my reply below, so you understand why I'm making this point. I do think you probably want to re-try in these cases. If it fails after three tries, it'll move on.

File: canvas/canvas_course_usage.py [hunk 16]
```diff
@@ -0,0 +1,107 @@
...
+                self.course_retry_list = []
+                logger.info("Retry round getting canvas_course_usage data")
+                responses = [session.get(f'{self.canvas_url[0]}/api/v1/courses/{retry_course}/analytics/activity',
+                                         headers=headers) for retry_course in retry_courses]
+
+            for response in as_completed(responses):
+                self.parsing_canvas_course_usage_data(response)
+
+        logger.info(f"Any thing to Retry? With List of length {len(self.course_retry_list)} : {self.course_retry_list}")
+        if len(self.course_retry_list) != 0 and self.retry_count < self.retry_attempts:
```
Comment (pushyamig) [inline canvas/canvas_course_usage.py:75]: Here I am checking if you have any more data to retry

**Comment by pushyamig** [comment]
Each time I retry I one try failed cases from last time
Round 1: Course1, Course2, Course3
Round 2: Course 2 ( try since it failed again), Course1, course3 (was able to get data from this try)
Round 3: Course 2 still has a problem, I exit at that time with this check `if len(self.course_retry_list) != 0 and self.retry_count < self.retry_attempts:`

make sense?

File: canvas/canvas_course_usage.py (hunk 7 above)
Comment (jonespm) [reply canvas/canvas_course_usage.py:26]: I never thought much about this before. In the [PEP-8](https://www.python.org/dev/peps/pep-0008/#programming-recommendations) it says "Be consistent in return statements. Either all return statements in a function should return an expression, or none of them should.". In this function it looks okay but I also like the explicit `return None` and have been trying to use it more and remember to check more for it. Either `return None` and `return` both still return None. 

FYI: If you were type hinting this putting `-> None: ` would make it clear that this only returns None.

File: canvas/canvas_course_usage.py (hunk 15 above)
Comment (ssciolla) [reply canvas/canvas_course_usage.py:19]: I thought you were incremented this for each time something failed, but you're only incrementing if one or more requests failed. I think that should be okay, another way of doing it.

File: inventory.py [hunk 17]
```diff
@@ -238,6 +244,10 @@ def run_course_inventory() -> None:
         enrollment_df.to_csv(os.path.join('data', 'enrollment.csv'), index=False)
         logger.info('Wrote data to data/enrollment.csv')
 
+        logger.info(f"Writing {num_canvas_usage_records} canvas course usage record to CSV")
```
Comment (ssciolla) [inline inventory.py:247]:
```suggestion
        logger.info(f"Writing {num_canvas_usage_records} canvas course usage records to CSV")
```

File: canvas/canvas_course_usage.py (hunk 10 above)
Comment (ssciolla) [reply canvas/canvas_course_usage.py]: Right when I figured out those commas were making those instance variables store tuples, I saw your commit. Thanks!

**Comment by pushyamig** [comment]
I have a commit for the canvas token indexing, which was annoying I did not even realize that I was declaring the canvas token/url as tuple 
https://stackoverflow.com/questions/27654671/member-variable-string-gets-treated-as-tuple-in-python

**Comment by pushyamig** [comment]
For the Return None , I am type hinting for now. Thanks for more insights on that

File: canvas/canvas_course_usage.py [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+                rows.append(row)
+
+        df = pd.DataFrame(rows)
+        logger.info(df.head())
+        df.drop(['id'], axis=1, inplace=True)
+        df_dup = df[df.duplicated()]
+        logger.info('Check for duplicate items ')
+        logger.info(df_dup)
+        df = df.drop_duplicates()
+        logger.info(df.head())
```
Comment (ssciolla) [inline canvas/canvas_course_usage.py:100]: @jonespm and I talked about the use of `inplace` a little bit, which I think will be deprecated in the future. See issue #52. Would you mind changing line 95 here to `df = df.drop(['id], axis=1)`. It's just a little more clear to me, and you're also not using `inplace` on line 99, so it'd be a little more consistent.

File: canvas/canvas_course_usage.py [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+        if not analytics_data:
+            logger.info(f"Response for fetching canvas course usage is empty")
+            return
+        new_dic['course_id'] = course_id
+        new_dic['analytics'] = analytics_data
+
+        self.canvas_usage_courses.append(new_dic)
+
+    def _get_canvas_course_views_participation_data(self, retry_courses=None):
+        logger.info("Starting of _get_canvas_course_views_participation_data call")
```
Comment (ssciolla) [inline canvas/canvas_course_usage.py:59]: Maybe we can make this `debug` level?

File: canvas/canvas_course_usage.py [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+        self.canvas_url = canvas_url
+        self.canvas_token = canvas_token
+        self.course_ids = course_ids
+        self.retry_attempts = retry_attempts
+        self.canvas_usage_courses = []
+        self.course_retry_list = []
+        self.retry_count = 0
+
+    def parsing_canvas_course_usage_data(self, response) -> None:
+        logger.info("parsing_canvas_course_usage_data Call")
```
Comment (ssciolla) [inline canvas/canvas_course_usage.py:22]: We could also potentially make this `debug`. There are just going to be a lot of these, and it seems like it would be most useful when changing code.

File: canvas/canvas_course_usage.py [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+
+            for row in analytics:
+                row['course_id'] = course_id
+                rows.append(row)
+
+        df = pd.DataFrame(rows)
+        logger.info(df.head())
+        df.drop(['id'], axis=1, inplace=True)
+        df_dup = df[df.duplicated()]
+        logger.info('Check for duplicate items ')
```
Comment (ssciolla) [inline canvas/canvas_course_usage.py:97]:
```suggestion
        logger.info('Check for duplicate items')
```

File: canvas/canvas_course_usage.py (hunk 18 above)
Comment (ssciolla) [inline canvas/canvas_course_usage.py:100]: `debug`?

**Comment by pushyamig** [comment]
@ssciolla , I think I have resolved your recent review suggestion as well

**Comment by ssciolla** [comment]
> @ssciolla , I think I have resolved your recent review suggestion as well

Thanks!

**Comment by pushyamig** [comment]
I want to run this Branch on test instance before I merge it master. I just want to see how the MYSQL test instance handles 500K row of date delete and insert on the AWS Mysql instance
